File: KSC2013_ProblemSet_n_Solution/2013년문제답안/Problem3/Solution/S3-parallel.f90
SUBROUTINE advance_field(nx, ny, f, g, c)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nx, ny
  REAL(KIND=8), DIMENSION(nx,ny), INTENT(IN)    :: g, c
  REAL(KIND=8), DIMENSION(nx,ny), INTENT(INOUT) :: f

  INTEGER :: i,j

  DO j=2,ny-1
    DO i=2,nx-1
      f(i,j) = c(i,j)*(g(i+1,j) + g(i-1,j) + g(i,j+1) + g(i,j-1) - 4*g(i,j)) &
          + 2*g(i,j) - f(i,j)
    END DO
  END DO
END SUBROUTINE




SUBROUTINE periodic_x(nx, ny, f)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nx, ny
  REAL(KIND=8), DIMENSION(nx,ny), INTENT(INOUT) :: f

  f(nx,:) = f(2,:)
  f(1,:) = f(nx-1,:)
END SUBROUTINE




SUBROUTINE exchange_boundary(nx, ny, f)
  IMPLICIT NONE
  INCLUDE 'mpif.h'
  INTEGER, INTENT(IN) :: nx, ny
  REAL(KIND=8), DIMENSION(nx,ny), INTENT(INOUT) :: f
  INTEGER :: ierr, nprocs, myrank
  INTEGER :: ireq1, ireq2, ireq3, ireq4, status(MPI_STATUS_SIZE)

  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)


  IF (myrank == 0) THEN
    CALL MPI_ISEND(f(1,ny-1), nx, MPI_REAL8, 1, 10, MPI_COMM_WORLD, ireq1, ierr)
    CALL MPI_IRECV(f(1,ny), nx, MPI_REAL8, 1, 20, MPI_COMM_WORLD, ireq2, ierr)

  ELSE IF (myrank == nprocs-1) THEN
    CALL MPI_ISEND(f(1,2), nx, MPI_REAL8, nprocs-2, 20, MPI_COMM_WORLD, ireq1, ierr)
    CALL MPI_IRECV(f(1,1), nx, MPI_REAL8, nprocs-2, 10, MPI_COMM_WORLD, ireq2, ierr)

  ELSE
    CALL MPI_ISEND(f(1,ny-1), nx, MPI_REAL8, myrank+1, 10, MPI_COMM_WORLD, ireq1, ierr)
    CALL MPI_ISEND(f(1,2), nx, MPI_REAL8, myrank-1, 20, MPI_COMM_WORLD, ireq2, ierr)
    CALL MPI_IRECV(f(1,ny), nx, MPI_REAL8, myrank+1, 20, MPI_COMM_WORLD, ireq3, ierr)
    CALL MPI_IRECV(f(1,1), nx, MPI_REAL8, myrank-1, 10, MPI_COMM_WORLD, ireq4, ierr)

    CALL MPI_WAIT(ireq3, status, ierr)
    CALL MPI_WAIT(ireq4, status, ierr)
  END IF

  CALL MPI_WAIT(ireq1, status, ierr)
  CALL MPI_WAIT(ireq2, status, ierr)
END SUBROUTINE




PROGRAM wave2d
  IMPLICIT NONE
  INCLUDE 'mpif.h'
  INTEGER, PARAMETER :: tnx=8000, tny=8000
  INTEGER, PARAMETER :: tmax=3000
  INTEGER, PARAMETER :: width=120, thick=60, gap=1200   ! slit parameters
  INTEGER, PARAMETER :: distance=1500
  REAL(KIND=8), PARAMETER :: c0=0.49D0
  REAL(KIND=8), ALLOCATABLE, DIMENSION(:,:) :: f, g, c
  REAL(KIND=8), DIMENSION(tnx,tny) :: fout
  INTEGER :: nx, ny
  INTEGER :: i,j,tstep
  INTEGER :: ierr, nprocs, myrank, rank


  !----------------------------------------------------------------------------
  ! initialize the MPI environmnet
  !----------------------------------------------------------------------------
  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)


  !----------------------------------------------------------------------------
  ! allocate the field arrays
  !----------------------------------------------------------------------------
  nx = tnx
  ny = tny/nprocs + 2
  ALLOCATE(f(nx,ny))
  ALLOCATE(g(nx,ny))
  ALLOCATE(c(nx,ny))


  !----------------------------------------------------------------------------
  ! initialize coefficient and fields
  !----------------------------------------------------------------------------
  DO j=1,ny
    DO i=1,nx
      f(i,j) = 0D0
      g(i,j) = 0D0
      c(i,j) = c0
    END DO
  END DO


  !----------------------------------------------------------------------------
  ! slit structure
  !----------------------------------------------------------------------------
  ! We assume the nprocs is multiple of two.
  IF (myrank == nprocs/2) THEN
    c(:,2:2+thick) = 0D0
    c(nx/2-(gap+width)/2:nx/2+(gap+width)/2,2:2+thick) = c0
    c(nx/2-(gap-width)/2:nx/2+(gap-width)/2,2:2+thick) = 0D0
  END IF
  

  !----------------------------------------------------------------------------
  ! main loop for the time evolution
  !----------------------------------------------------------------------------
  DO tstep=1,tmax
    !------------------------------------
    ! point source
    !------------------------------------
    ! IF (myrank == 0) THEN
    !   g(200,ny-200) = SIN(0.02*tstep)
    ! END IF


    !------------------------------------
    ! line source
    !------------------------------------
    ! We assume the nprocs is multiple of two.
    IF (myrank == nprocs/2 - distance/ny - 1) THEN
      g(:,ny-(distance-distance/ny*ny)) = SIN(0.02*tstep)
    END IF


    CALL advance_field(nx, ny, f, g, c)
    CALL periodic_x(nx, ny, f)
    CALL exchange_boundary(nx, ny, f)

    CALL advance_field(nx, ny, g, f, c)
    CALL periodic_x(nx, ny, g)
    CALL exchange_boundary(nx, ny, g)
  END DO


  !----------------------------------------------------------------------------
  ! gather fields and save as binary files
  !----------------------------------------------------------------------------
  CALL MPI_GATHER(f(1,2), nx*(ny-2), MPI_REAL8, fout, nx*(ny-2), MPI_REAL8, 0, MPI_COMM_WORLD, ierr)

  IF (myrank == 0) THEN
    OPEN(UNIT=11, FILE='field.bin', ACCESS='DIRECT', FORM='UNFORMATTED', &
        RECL=tnx*tny*8, STATUS='NEW')
    WRITE(11,rec=1) fout
    CLOSE(11)
  END IF


  !----------------------------------------------------------------------------
  ! finalize the MPI environmnet
  !----------------------------------------------------------------------------
  CALL MPI_FINALIZE(ierr)
END PROGRAM

File: KSC2013_ProblemSet_n_Solution/2013년문제답안/Problem3/Solution/S3-parallel.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<mpi.h>



void advance_field(int nx, int ny, double *f, double *g, double *c) {
	int i, j, idx;

	for (i=1; i<nx-1; i++) {
		for (j=1; j<ny-1; j++) {
			idx = i*ny + j;

			f[idx] = c[idx]*(g[idx+ny] + g[idx-ny] + g[idx+1] + g[idx-1] - 4*g[idx]) 
		             + 2*g[idx] - f[idx];
		}
	}
}



void periodic_y(int nx, int ny, double *f) {
	int i, j;

	for (i=0; i<nx; i++) {
		f[i*ny+(ny-1)] = f[i*ny+1];
		f[i*ny+0] = f[i*ny+(ny-2)];
	}
}



void exchange_boundary(int nx, int ny, double *f) {
	int i, j;
	int nprocs, myrank;
	MPI_Request req1, req2, req3, req4;
	MPI_Status stat;

	MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
	MPI_Comm_rank(MPI_COMM_WORLD, &myrank);


	if (myrank < nprocs-1) {
		MPI_Isend(&f[(nx-2)*ny], ny, MPI_DOUBLE, myrank+1, 10, MPI_COMM_WORLD, &req1);
		MPI_Irecv(&f[(nx-1)*ny], ny, MPI_DOUBLE, myrank+1, 20, MPI_COMM_WORLD, &req2);
	}

	if (myrank > 0 ) {
		MPI_Isend(&f[ny], ny, MPI_DOUBLE, myrank-1, 20, MPI_COMM_WORLD, &req3);
		MPI_Irecv(&f[0], ny, MPI_DOUBLE, myrank-1, 10, MPI_COMM_WORLD, &req4);
	}

	if (myrank < nprocs-1) {
		MPI_Wait(&req1, &stat);
		MPI_Wait(&req2, &stat);
	}

	if (myrank > 0 ) {
		MPI_Wait(&req3, &stat);
		MPI_Wait(&req4, &stat);
	}
}



int main(int argc, char *argv[]) {
	int tnx=8000, tny=8000;
	int tmax=3000;
	int width=120, thick=60, gap=1200;	// slit parameters
	int distance=1500;
	double c0=0.49;
	double *f, *g, *c, *ftot;
	int nx, ny;
	int i, j, idx, tstep;
	int nprocs, myrank;


    //-------------------------------------------------------------------------
    // initialize the MPI environment
    //-------------------------------------------------------------------------
	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
	MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
	

	
    //-------------------------------------------------------------------------
    // initialize coefficient and fields
    //-------------------------------------------------------------------------
	nx = tnx/nprocs + 2;
	ny = tny;

	f = (double*)malloc(nx*ny*sizeof(double));
	g = (double*)malloc(nx*ny*sizeof(double));
	c = (double*)malloc(nx*ny*sizeof(double));

	for (i=0; i<nx*ny; i++) {
		f[i] = 0;
		g[i] = 0;
		c[i] = c0;
	}
	

    //-------------------------------------------------------------------------
  	// slit structure
    //-------------------------------------------------------------------------
    // We assume the nprocs is multiple of two.
	if (myrank == nprocs/2) {
		for (i=1; i<thick+2; i++) {
			for (j=0; j<ny; j++)
				c[i*ny+j] = 0;
		}

		for (i=1; i<thick+2; i++) {
			for (j=ny/2-(gap+width)/2-1; j<ny/2+(gap+width)/2; j++)
				c[i*ny+j] = c0;
		}
		
		for (i=1; i<thick+2; i++) {
			for (j=ny/2-(gap-width)/2-1; j<ny/2+(gap-width)/2; j++)
				c[i*ny+j] = 0;
		}
	}


    //-------------------------------------------------------------------------
  	// main loop for the time evolution
    //-------------------------------------------------------------------------
	for (tstep=1; tstep<=tmax; tstep++) {
    	//------------------------------------
    	// point source
    	//------------------------------------
		//if (myrank == 0) 
	    //    g[nx/2-1][ny/2-1] = sin(0.02*tstep);


    	//------------------------------------
    	// line source
    	//------------------------------------
        // We assume the nprocs is multiple of two.
		if (myrank == nprocs/2 - distance/nx - 1) 
			for (j=0; j<ny; j++)
				g[(nx-(distance-distance/nx*nx))*ny+j] = sin(0.02*tstep);


		advance_field(nx, ny, f, g, c);
		periodic_y(nx, ny, f);
    	exchange_boundary(nx, ny, f);

		advance_field(nx, ny, g, f, c);
		periodic_y(nx, ny, g);
    	exchange_boundary(nx, ny, g);
	}


    //-------------------------------------------------------------------------
  	// gather fields and save as binary files
    //-------------------------------------------------------------------------
	ftot = (double*)malloc(tnx*tny*sizeof(double));
  	MPI_Gather(&f[1], (nx-2)*ny, MPI_DOUBLE, ftot, (nx-2)*ny, MPI_DOUBLE, 0, MPI_COMM_WORLD);

	if (myrank == 0) {
		FILE *fout;
		fout = fopen("field.bin", "wb");
		fwrite(ftot, sizeof(double), tnx*tny, fout);
		fclose(fout);
	}


    //-------------------------------------------------------------------------
    // finalize the MPI environment
    //-------------------------------------------------------------------------
	MPI_Finalize();

	return 0;
}
